### PR TITLE
gh-124606: Fix reference leak in error path in `datetime_fromisoformat` in `_datetimemodule.c`

### PR DESCRIPTION
**Pull request by Eclips4**
* Issue: gh-124606

### CODE REVIEW

**Review by picnixz — APPROVED** [review]
Thank you for the fix! (though I find the gotos a bit too hard to follow in this specific function). Do you want to run the refleak build bots?

**Review by vstinner — APPROVED** [review]
LGTM.

I confirm that `./python -m test test_datetime -R 3:3` no longer leaks with this change.

For the PR title, you may refer to datetime_fromisoformat() which is more precise ;-)

**Comment by vstinner** [comment]
> Do you want to run the refleak build bots?

It's not needed. I tested manually.
